Fixed panic recovery loop #2

dennis-tra opened the pull request:
Another approach to the problem outlined in PR #334:

We're relying on the cluster provider (consul in our case) to reflect a consistent state of the cluster topology. If a node is still part of the cluster (according to consul) we accept the panic/recovery loop, because it could be that it is just temporarily not available. As soon as the cluster provider says that the node has left the cluster the connection attempts are stopped by the `connectionDecider` - a new field on the `remoteConfig`.

In our case the `connectionDecider` checks if the address we're trying to connect to is still part of the cluster. If it is not we stop gracefully.
```go
remotingOption = append(remotingOption, remote.WithConnectionDecider(func(address string) bool {
	adrList := cluster.GetMemberAddresses()
	for _, v := range adrList {
		if v == address {
			return true
		}
	}
	log.Infof("Address %s is no known member. Deciding not to connect", address)
	return false
}))
```

Remarks: 
I would prefer if this logic didn't need to be put in from the outside but rather be handled internally. But we can't have a dependency from the `remote` to the `cluster` package -- This could be solved when the default `ClusterConfig` would have the `connectionDecider` `RemotingOption` already attached to it.

Strum355 commented on cluster/cluster.go:
```diff
@@ -122,7 +122,23 @@ func GetMemberPIDs(kind string) actor.PIDSet {
 
-// RemoveCache at PidCache
+// GetMemberAddresses returns the list of addresses of all currently known cluster members.
+func GetMemberAddresses() []string {
```
Im curious why this function is here if its not being used anywhere else in the PR

dennis-tra replied:
Good catch, this is used inside the `connectionDecider` function. I forgot to put it in here - but you can find it's usage in your PR #334. Here the excerpt:

```go
remotingOption = append(remotingOption, remote.WithConnectionDecider(func(address string) bool {
	adrList := cluster.GetMemberAddresses()
	for _, v := range adrList {
		if v == address {
			return true
		}
	}
	log.Infof("Address %s is no known member. Deciding not to connect", address)
	return false
}))
```

I'll update the PR

Strum355 commented:
We're going to tryout this fix right now and will report back :)

dennis-tra commented:
Oh cool 👍 yeah let me know! :)

dennis-tra commented:
Oh hang on, rather try the other one :D #350 that's the one I prefer

Strum355 commented:
Trying to figure out how to repro the issue we're having first :D this one was easier to add quickly to the code so it got preference ;)

fabotronix committed:
Add test for possible deadlock

fabotronix committed:
Fix possible deadlock

When the memberlist publishes events on the event stream while beeing triggered
by an event from the event stream itself, the event stream read lock is
acquired twice. This may lead to a deadlock, when in parallel a write
lock for the event stream is acquired. (e.g. due to a Subscribe or Unsubscribe)

goroutine 179 [semacquire, 25 minutes]:
sync.runtime_SemacquireMutex(0x193226c, 0xc00016e000, 0x0)
	/usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:50
github.com/AsynkronIT/protoactor-go/eventstream.(*EventStream).Publish(0x1932260, 0xef6060, 0xc025126270)
	/go/src/github.com/AsynkronIT/protoactor-go/eventstream/eventstream.go:65 +0x161
github.com/AsynkronIT/protoactor-go/eventstream.Publish(...)
	/go/src/github.com/AsynkronIT/protoactor-go/eventstream/eventstream.go:21
github.com/AsynkronIT/protoactor-go/cluster.(*memberListValue).updateAndNotify(0xc00079b640, 0xc01b7640c0, 0xc02584ba40)
	/go/src/github.com/AsynkronIT/protoactor-go/cluster/member_list.go:193 +0x4c9
github.com/AsynkronIT/protoactor-go/cluster.(*memberListValue).updateClusterTopology(0xc00079b640, 0xe9ca00, 0xc01c927300)
	/go/src/github.com/AsynkronIT/protoactor-go/cluster/member_list.go:105 +0x364
github.com/AsynkronIT/protoactor-go/eventstream.(*EventStream).Publish(0x1932260, 0xe9ca00, 0xc01c927300)
	/go/src/github.com/AsynkronIT/protoactor-go/eventstream/eventstream.go:70 +0xd4
github.com/AsynkronIT/protoactor-go/eventstream.Publish(...)
	/go/src/github.com/AsynkronIT/protoactor-go/eventstream/eventstream.go:21
github.com/AsynkronIT/protoactor-go/cluster/consul.(*ConsulProvider).notifyStatuses(0xc000208540)
	/go/src/github.com/AsynkronIT/protoactor-go/cluster/consul/consul_provider.go:276 +0x99f
github.com/AsynkronIT/protoactor-go/cluster/consul.(*ConsulProvider).MonitorMemberStatusChanges.func1(0xc000208540)
	/go/src/github.com/AsynkronIT/protoactor-go/cluster/consul/consul_provider.go:282 +0x28
created by github.com/AsynkronIT/protoactor-go/cluster/consul.(*ConsulProvider).MonitorMemberStatusChanges
	/go/src/github.com/AsynkronIT/protoactor-go/cluster/consul/consul_provider.go:280 +0x3f
